Some FSI additions

lindsayad opened the pull request:
- Interface kernel for coupling scalar field displacments to vector fluid velocity field
- Create ConvectedMeshPSPG so that we can have one-to-one comparison between INS and INSAD in the FSI module
- Fix mesh convection residual for INSAD. The multiplication order needs to be switched to match the regular advection term

Closes #23680

lindsayad commented:
@eilloo I think the work here is relevant to some of your discussion posts as well as #23680

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1487475/) on 8c5f68c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24204/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1495723/) on 8a8ed05 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c5d416</th>
      <th colspan=4>#24204 8a8ed0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.17%</td>
      <td>85.17%</td>
      <td>+0.00%</td>
      <td>75.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87106</td>
      <td>87087</td>
      <td>-19</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15169</td>
      <td>15164</td>
      <td>-5</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>c5d416</th>
      <th colspan=4>#24204 8a8ed0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.13%</td>
      <td>90.13%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3535</td>
      <td>3533</td>
      <td>-2</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>387</td>
      <td>387</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage/contact)

### Fsi

<table>
  <thead>
    <tr>
      <th></th>
      <th>c5d416</th>
      <th colspan=4>#24204 8a8ed0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.48%</td>
      <td>89.26%</td>
      <td>-1.22%</td>
      <td>88.04%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>171</td>
      <td>324</td>
      <td>+153</td>
      <td>162</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18</td>
      <td>39</td>
      <td>+21</td>
      <td>22</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage_diff/fsi)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage/fsi)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>c5d416</th>
      <th colspan=4>#24204 8a8ed0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.53%</td>
      <td>83.12%</td>
      <td>-0.41%</td>
      <td>54.75%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10688</td>
      <td>10742</td>
      <td>+54</td>
      <td>98</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2107</td>
      <td>2181</td>
      <td>+74</td>
      <td>81</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage/navier_stokes)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>c5d416</th>
      <th colspan=4>#24204 8a8ed0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.77%</td>
      <td>93.77%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4080</td>
      <td>4079</td>
      <td>-1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>271</td>
      <td>271</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24204/coverage/ray_tracing)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24204/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 75.00% is less than the suggested 90.0%
- `fsi` new line coverage rate 88.04% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 54.75% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1495723/) on 8a8ed05 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.26% is less than the required 90.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1548796/) on b617523 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.23% is less than the required 90.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1551942/) on c52fa14 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 82.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)

moosebuild commented:
All [jobs](https://civet.inl.gov/event/126710/) on c52fa14 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1551942/) on c52fa14 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.23% is less than the required 90.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1555564/) on de799b1 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.23% is less than the required 90.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1555959/) on 082c518 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.23% is less than the required 90.0%

lindsayad commented:
For ease of reviewing I'm going to split this into smaller chunks starting with #24514 and #24515

lindsayad commented:
With #24515 merged, turning to #24540

lindsayad commented:
Once the NS PR goes in then I will return here for the FSI code

lindsayad committed:
FSI additions

- Interface kernel for coupling scalar field displacments to vector
  fluid velocity field
- Create ConvectedMeshPSPG so that we can have one-to-one comparison
  between INS and INSAD in the FSI module
- Fix mesh convection residual for INSAD. The multiplication order needs
  to be switched to match the regular advection term

Closes #23680

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1759700/) on d3f0016 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.23% is less than the required 90.0%

lindsayad committed:
Make FSI coverage requirement less strict

maxnezdyur reviewed (commented):
Looking at how MeshConvection was implemented for INSAD, everything looks correct, but I think the relative velocity would need to be used in place of the fluid velocity for a consistent stabilization. I don't know how important that is here but wanted to make you aware of it.

lindsayad commented:
> but I think the relative velocity would need to be used in place of the fluid velocity for a consistent stabilization.

Looking at [this 2019 article](https://www.researchgate.net/profile/Tayfun-Tezduyar/publication/331353222_A_Stabilized_ALE_Method_for_Computational_Fluid-Structure_Interaction_Analysis_of_Passive_Morphing_in_Turbomachinery/links/5cfc6fe94585157d159e089a/A-Stabilized-ALE-Method-for-Computational-Fluid-Structure-Interaction-Analysis-of-Passive-Morphing-in-Turbomachinery.pdf) I do see the modification to the velocity used to multiply the test function gradient in the SUPG term but I do not see it used in tau. However, that seems like an oversight to me. The momentum advection will be defined by the relative velocity (e.g. u - u_hat) so it makes sense to me that if the relative velocity is 0, then the contribution to tau from advection should be 0

maxnezdyur commented:
I thought the advection-dominated limit term would also take into account the mesh velocity, but I can't seem to find my source, so I may be mistaken there. Basically, I would expect everywhere we use `U.norm()` in our tau parameters we would use `(U-U_hat).norm()`. I'll try to find my source though.

lindsayad commented:
Source would be great, but yea I agree that seems logical

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/1805847/) on 308ae5b : invalidated by @lindsayad

lindsayad committed:
More convected -> advected language change

lindsayad committed:
Use relative velocity for SUPG velocity and tau computation

lindsayad committed:
Modify INS to use relative velocity in SUPG/tau computations

INS and INSAD now match again

lindsayad committed:
Remove INS Jacobian test

Using the relative velocity as the advecting velocity in SUPG and
in calculations of tau introduced new dependence on the displacements
that I don't want to spend time coding the Jacobians for

lindsayad commented:
This now sits on top of https://github.com/idaholab/moose/pull/25574 so probably should wait for further review until that is merged

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1815702/) on 04d15cb : invalidated by @lindsayad

lindsayad committed:
Try to address Max's markdown comment

lindsayad commented:
@maxnezdyur has reviewed this. @loganharbour, @GiudGiud, or @roystgnr could I get a CCB review when one of you has a chance?

GiudGiud commented on modules/navier_stokes/include/materials/INSADTauMaterial.h:
```diff
@@ -90,11 +90,15 @@ class INSADTauMaterialTempl : public T
   const unsigned int _vel_sys_number;
 
+  /// The speed of the medium. This is the norm of the relative velocity, e.g. the velocity minus
+  /// the mesh velocity
```
```suggestion
  /// the mesh velocity, at the current _qp
```

GiudGiud commented on modules/navier_stokes/include/kernels/INSBase.h:
```diff
@@ -106,4 +110,13 @@ class INSBase : public Kernel
+
+  /// Whether displacements are provided
+  const bool _disps_provided;
+  /// Time derivative of the x-displacement
```
```suggestion
  /// Time derivative of the x-displacement, mesh velocity in the x-direction
```

GiudGiud commented on modules/fsi/test/tests/2d-small-strain-transient/tests:
```diff
@@ -1,14 +1,42 @@
+      exodiff = 'fsi_flat_channel_out.e'
+      detail = 'scalar field variables for velocity components and hand-coded Jacobians,'
+      # vel_x_solid is not accurate
+      # We mmay improve that in the future
```
```suggestion
      # We may improve that in the future
```

GiudGiud commented on modules/fsi/include/kernels/ConvectedMeshPSPG.h:
```diff
@@ -0,0 +1,48 @@
+#include "INSBase.h"
+
+/**
+ * This calculates the time derivative for a coupled variable
```
```suggestion
 *  Computes residual and Jacobian contributions for the PSPG stabilization term for mesh advection
```

GiudGiud commented on modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:
```diff
@@ -0,0 +1,149 @@
+    for (auto & r : _residuals)
+      r = 0;
+
+    for (const auto qp : make_range(_qrule->n_points()))
+      for (const auto i : index_range(phi))
+        _residuals[i] += phi[i][qp] * _qp_jumps[qp];
```
I would expect flipping the loops to be faster

lindsayad replied:
why?

GiudGiud replied:
more local? you re jumping around in memory for phi array every i (i * qp times) here.

the other one you loop on the inner index and you only jump in memory for phi i times

GiudGiud replied:
probably doesnt matter.

lindsayad replied:
change

GiudGiud commented on modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:
```diff
@@ -0,0 +1,149 @@
+    computeResidual();
+}
+
+void
+ADPenaltyVelocityContinuity::computeNeighborOffDiagJacobian(unsigned int)
+{
+}
+
+const MooseVariableFieldBase &
+ADPenaltyVelocityContinuity::variable() const
+{
+  return *_velocity_var;
+}
```
header?

lindsayad committed:
Apply suggestions from code review

- Better doxygen for ConvectedMeshPSPG
- Typo fix
- Better doxygen for _speed member
- Inline short methods in the header
- More doxygen for disp dot members in INSBase
- Switch i and qp loop order

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
thanks for the reviews!
